form feature/checkbox label for version

Chun-Lin opened the pull request:
## Done
#37 checkbox

1. Checkbox

> users can just use Checkbox component to build a checkbox with text

```js
<Checkbox>Test</Checkbox>
```

2. CheckboxGroup
 
> users can use CheckboxGroup component to build a list of checkboxes with options, also can assign direction to make either Row and Column side checkboxes

```js
<div>
  <CheckboxGroup
    options={['SUB', 'MNR', 'ICX']}
    span={30}
    theme="light"
    direction="row"
  />
</div>
```


### More details can be found in Styleguidist. 
```js
yarn styleguide
```

![fireshot capture 2 - checkbox style guide - http___localhost_6060_](https://user-images.githubusercontent.com/12542369/35895404-3c5632e4-0bf2-11e8-8f51-99ef3e84ac71.png)

Chun-Lin committed:
feat: add recompose and style-components

Chun-Lin committed:
feat: now checkbox can change state (checked)

checkbox can change its state when the user click the checkbox

Chun-Lin committed:
feat: add label next to the checkbox

add <label> tag in App.js to show the text and remove <div> in checkbox.js to make the text right
next to checkbox

Chun-Lin committed:
feat: add children as text input

add <label> in Checkbox.js as children input and remove <label> in app.js

Chun-Lin committed:
feat: add styles to checkbox

add a util function to load multiple icons as variables in js. Also using styled-components to style
Input and Span

Chun-Lin committed:
style: add font style to text

add font color style when checkbox is disabled and make 4 checkboxes in app.js

Chun-Lin committed:
feat(add onChange props to the component): add onChange props to the component, but onChange functio

Chun-Lin committed:
feat: use label for to dispaly input form

using label for to display input form and make text clickable

Chun-Lin committed:
refactor: write demo code on Checkbox.md and remove useless code on Checkbox.js

wirte the demo styleguide on Checkbox.md and change input type into checkbox, also change
defaultValue props to defaultChecked and remove value props

Chun-Lin committed:
feat: add dark them and CheckboxGroup component

now you can just use CheckboxGroup component to group multiple checkboxes. Also, dark theme is
available now

Chun-Lin committed:
feat: add cursor style

cursor not-allowed for disabled checkbox, and cursor pointer for normal checkbox

Chun-Lin committed:
refactor: move <img> src to function and edit the dark theme markdown

move <img> src to imgSrc function to make render function clean, and edit the dark theme markdwon
text because the input id is conflict with light theme (same text)

Chun-Lin committed:
feat: add direction to checkboxgroup component

now you can assing directino props to checkboxgroup to set the direction of checkboxgroup

Chun-Lin committed:
feat: add span feature for column direction

now can set span pixels when the checkboxgroup direction props set to column

Chun-Lin committed:
refactor: refactor some elements in render function to styled-components attributes

move img src and alts attrs to styled-components attrs options

Chun-Lin committed:
fix: fix margin-bottom and margin-right may conflict when using column and row checkboxgroup in the

added statement which direction the checkboxgroup is using, also add 'col' alternative to 'column'

Chun-Lin committed:
feat: new styled-components version and some style changed

add 3.1.0 styled-components and add #ffffff color for normal checkbox in dark theme

Chun-Lin committed:
feat: add proptypes via setPropTypes function

Chun-Lin committed:
feat: add __test__ directory and make a simple test for checkbox.js

add enzyme and raf module to run test and modify npm test script for jest enviromnent, also make a
simple test to test the children text is right

Chun-Lin committed:
docs: add jest coverage script to package.json

Chun-Lin committed:
test: add a test for testing render correct image

Chun-Lin committed:
test: add several snapshots tests to <Checkbox> and <CheckboxGroup>

Chun-Lin committed:
test: add checkboxGroup length test

Chun-Lin committed:
feat: add checked and disabled props to <checkboxGroup>

Chun-Lin committed:
fix: fix the proptypes don't show up on the styleguidist

remove setPropTypes and setDefaultProps recompose function and declare propTypes and defaultProps
outside checkbox function

Chun-Lin committed:
test: split <checkboxGroup> component tests to checkboxgroup.spec.js

Chun-Lin committed:
refactor(just a refacotr to package.json):

Chun-Lin commented on packages/checkbox/src/react-component-node-modules/components/Checkbox/Checkbox.js:
```diff
@@ -0,0 +1,137 @@
+        ? checked
+          ? btn_checkbox_presseddisable_black
+          : btn_checkbox_disable_black
+        : checked ? btn_checkbox_pressed_black : btn_checkbox_black,
```
有沒有更好的寫法？

purepennons replied:
每種顏色的圖片會有四種狀態，先將 import 的圖片整理成物件的形式。然後另外寫一支 getChecboxStatus 的 function，這樣閱讀性比較高。
e.g.
```js
const checkBoxState = {
  light: {
    0: ..., // no checked, no disabled
    1: ..., // no checked, disabled
    2: ...,  // checked, no disabled
    3: ...,  // checked, disabled
  },
  dark: {
    0: ..., // no checked, no disabled
    1: ..., // no checked, disabled
    2: ...,  // checked, no disabled
    3: ...,  // checked, disabled
  }
}

const getCheckboxStatus = ({ checked, disabled }) => parseInt(checked << 1, 10) + parseInt(disabled, 10)

<Img src={checkBoxState[theme][getCheckboxStatus(this.props)]} />
```

Chun-Lin commented on packages/checkbox/src/react-component-node-modules/components/Checkbox/Checkbox.js:
```diff
@@ -0,0 +1,137 @@
+  )
+}
+
+Checkbox.propTypes = {
```
recompose setPropTypes沒辦法呈現proptypes在styleguidist上，看網路上別人也有一樣的issue，還不知道是什麼原因，所以另外拉出來寫

purepennons replied:
這個我之前有遇到，所以其實我不會去用 setPropTypes。而且本身 propTypes 有文件的用意，硬寫進 compose 中反而閱讀性較差。

purepennons commented:
@Chun-Lin 

名稱 rename 成 `@react-qui/form`。然後 `package.json` 補上
```js
  "private": false,
  "publishConfig": {
    "access": "public"
  },
```
[https://github.com/qeek-dev/react-qui/blob/master/README.md#package-naming-in-packagejson](https://github.com/qeek-dev/react-qui/blob/master/README.md#package-naming-in-packagejson)

P.S: 我剛剛突然發現 `toastr` 沒有加上這段，所以發佈失敗了 XD... 有空也補一下。

purepennons commented on packages/checkbox/package.json:
```diff
@@ -0,0 +1,50 @@
+    "react": "^16.2.0",
+    "react-dom": "^16.2.0",
+    "recompose": "^0.26.0",
+    "styled-components": "^3.1.0"
```
除了 recompose 是相依 library 要保留，其他需要踢到 devDependencies 與 peerDependencies。([https://www.styled-components.com/docs/faqs#i-am-a-library-author-should-i-bundle-styledcomponents-with-my-library](https://www.styled-components.com/docs/faqs#i-am-a-library-author-should-i-bundle-styledcomponents-with-my-library))

# devDependencies
- akiya-react-component-scripts
- babel-plugin-syntax-dynamic-import
  - 這幹嘛的？
- enzyme-to-json
- react
- react-dom
- styled-components

# peerDependencies
- react
- react-dom
- styled-components

purepennons commented on packages/checkbox/package.json:
```diff
@@ -0,0 +1,50 @@
+{
+  "name": "checkbox",
```
[https://github.com/qeek-dev/react-qui/blob/master/README.md#package-naming-in-packagejson](https://github.com/qeek-dev/react-qui/blob/master/README.md#package-naming-in-packagejson)

purepennons commented on packages/checkbox/src/react-component-node-modules/components/Checkbox/Checkbox.js:
```diff
@@ -0,0 +1,137 @@
+  display: none;
+`
+
+const withContainer = compose(
```
對於 compose 後的 function，我們習慣稱呼為 `enhancer`

purepennons commented on packages/checkbox/src/react-component-node-modules/components/Checkbox/utils/utils.js:
```diff
@@ -0,0 +1,9 @@
+import pathParse from 'path-parse'
+
+export const importAllFiles = (ctx, cb) => ctx.keys().reduce((acc, curr) => {
```
沒用到就刪掉吧 XD

purepennons commented on packages/checkbox/src/react-component-node-modules/components/Checkbox/Checkbox.js:
```diff
@@ -0,0 +1,137 @@
+          ? btn_checkbox_presseddisable_black
+          : btn_checkbox_disable_black
+        : checked ? btn_checkbox_pressed_black : btn_checkbox_black,
+  alt: 'Src Error',
```
alt 的作用是在圖片失效的時候，給予本身此圖片的描述，使操作者仍然知道其意義。所以描述 Src Error 是無意義的。

purepennons commented on packages/checkbox/src/react-component-node-modules/components/Checkbox/Checkbox.js:
```diff
@@ -0,0 +1,137 @@
+`
+
+const withContainer = compose(
+  mapProps(props => ({
```
`mapProps` 我個人是有要預處理傳入的 props 才會使用，你這邊使用沒有意義，只是重複撰寫。

purepennons commented on packages/checkbox/src/react-component-node-modules/components/Checkbox/Checkbox.js:
```diff
@@ -0,0 +1,137 @@
+    theme: props.theme,
+    direction: props.direction,
+  })),
+  withState('checked', 'setChecked', props => props.checked),
```
`withState` + `withHandlers` 可以直接用 [`withStateHandlers`](https://github.com/acdlite/recompose/blob/master/docs/API.md#withstatehandlers) 取代。

purepennons commented on packages/checkbox/src/react-component-node-modules/components/Checkbox/Checkbox.js:
```diff
@@ -0,0 +1,137 @@
+  })),
+  withState('checked', 'setChecked', props => props.checked),
+  withHandlers({
+    handleClick: ({ setChecked, disabled }) => props => {
```
`withHandlers` 裡面的 method 的形式是 
```js
withHandlers(
  handlerCreators: {
    [handlerName: string]: (props: Object) => Function
  } |
  handlerCreatorsFactory: (initialProps) => {
    [handlerName: string]: (props: Object) => Function
  }
): HigherOrderComponent
```
所以應該是 ({ setChecked, disabled }) => event => {}

[https://github.com/acdlite/recompose/blob/master/docs/API.md#withhandlers](https://github.com/acdlite/recompose/blob/master/docs/API.md#withhandlers)

purepennons commented on packages/checkbox/src/react-component-node-modules/components/Checkbox/Checkbox.js:
```diff
@@ -0,0 +1,137 @@
+  disabled,
+  className,
+  onChange,
+  onClick,
```
像 onChange/onClick 這類 props，雖然已經透過 propTypes 指定 type 為 func，但是沒有初始值，這可能會導致 function 沒參數傳入時是空值，執行後程式 crash。我的作法會用 lodash 的 [noop](https://lodash.com/docs/4.17.4#noop) function 來當做初始化的 function。

```js
import { noop } from 'lodash/core'
const ({ onClick = noop }) => {...}
```

purepennons commented on packages/checkbox/src/react-component-node-modules/components/Checkbox/Checkbox.js:
```diff
@@ -0,0 +1,137 @@
+    >
+      <Img theme={theme} disabled={disabled} checked={checked} />
+      <Input
+        id={children}
```
children 可能是 object，把它當作 id 很怪（且可能壞掉），隨便用 shortid, nanoid 或 lodash 的 uniqueId 產生 id 都好。不過這邊的情況，因為你已經直接用 label 夾住 input 了，所以其實不需要在用 htmlFor 去綁定 input，他會自動綁定。需要綁定的時候是 label 與 input 非父子關係的時候。

purepennons commented on packages/checkbox/src/react-component-node-modules/components/Checkbox/Checkbox.js:
```diff
@@ -0,0 +1,137 @@
+  color: ${props =>
+    props.theme === 'light'
+      ? props.disabled ? '#aeaeae' : null
+      : props.disabled ? '#aeaeae' : '#ffffff'};
```
老樣子，色碼可以獨立拉出來。
然後 styled-component 中不要 return null，而是要給定一個完整的預設值，否則屬性數值會直接是 null。

[https://github.com/styled-components/styled-components/issues/275](https://github.com/styled-components/styled-components/issues/275)

purepennons replied:
[https://github.com/styled-components/styled-components/blob/master/docs/tips-and-tricks.md](https://github.com/styled-components/styled-components/blob/master/docs/tips-and-tricks.md)

Chun-Lin committed:
fix: move some dependencies to devDependencies and peerDependencies

move akyia-react-componenent-scripts, enzyme-to-json, react, react-dom, styled-components to
devDepenedencies and react, react-dom, styled-components to peerDependencies

Chun-Lin committed:
fix: delete useless utils]

Chun-Lin committed:
fix: make alt response more resonable and remove mapProps function

Chun-Lin committed:
style: make checkbox img more centered next to text and also fix the verical span between checkboxes

Chun-Lin committed:
refactor: integrate withState and withHandler to withStateHandler

Chun-Lin committed:
fix: add noop function from lodash/fp and attach it on onClick and onChange

Chun-Lin committed:
fix: add shortid generator to input id and make sure styled-components won't return null

Chun-Lin committed:
test: update snapshots

Chun-Lin committed:
feat: add onChange event to <CheckboxGroup> and make options can be checked or disabled manually

add onChange event to <CheckboxGroup>, now users can trigger onChange function via <CheckboxGroup>.
Also, checked and disabled features can be added manually to each option

Chun-Lin committed:
fix: move shortid to devDependencies and peerDependencies

Chun-Lin committed:
refactor: make img src scope more readable

move img src and alts to objects and give every images symbols to make it more readable

Chun-Lin committed:
test: modify some test cases to match the props of options in <CheckboxGroup>

Commit:
fix: seperate value props from children, and now text would be label in the options props

CheckboxGroup children is now label element in options, and value can be assinged by users

Commit:
style: move color to colors.js

Commit:
test: modify test suites based-on changed things from checkboxgroup options

Commit:
fix: export colors.js and modify markdown files

Chun-Lin committed:
fix: make checkbox a controlled input component

now Checkbox is a controlled component, you can controll the state outside it

Chun-Lin committed:
refactor: refactor CheckboxGroup

now Checkbox can be the children inside CheckboxGroup, so options props are removed. Users can
manage checked props inside every Chekcbox component

Chun-Lin committed:
fix: fix a single checkbox state cannot be changed at first time

Chun-Lin committed:
style: add rem functino to all styles which need pixels on their elements

Chun-Lin committed:
refactor: remove some useless props and add some missing props

Chun-Lin committed:
refactor: move <Input> out of <Label> and hook it with htmlFor

Chun-Lin committed:
refactor: move Checkbox component into form directory

Chun-Lin committed:
refactor: modify the chidren propTypes to PropTypes.node

Chun-Lin committed:
test: modify test suite

Chun-Lin commented:
local測試pass，但是circle CI測試沒過，檢查過snapshot應該是沒問題，不確定什麼原因

Chun-Lin committed:
test: change snapshot mothed from shallow to mount
